src/goto-symex/field_sensitivity.cpp

```diff
@@ -97,10 +97,21 @@ exprt field_sensitivityt::apply(
       // SSA expression
       ssa_exprt tmp = to_ssa_expr(index.array());
       bool was_l2 = !tmp.get_level_2().empty();
+
+      exprt array_size_expr = to_array_type(l2_array.get().type()).size();
+      if(array_size_expr.is_nil() && index.array().id() == ID_symbol)
+      {
+        // In case the array type was incomplete, attempt to retrieve it from
+        // the symbol table.
+        const symbolt *array_from_symbol_table = ns.get_symbol_table().lookup(
```

Comment:
What is the justification for this to be correct? I don't see why we could use values from the symbol table during symbolic execution.

Reply:
I've only seen that happen in one case which is this regression test: https://github.com/diffblue/cbmc/blob/develop/regression/cbmc/Global_Initialization2/main.c
In this example since `A1` is declared but not defined the length we get at this point in symex is not defined, although it appears correctly in the symbol table. @peterschrammel seemed to think that could be caused by a bug in the C frontend but we couldn't identify the source of this issue.

Reply:
As far as I remember the problem is that the array type is created with undefined length and then attached to various expressions from which the goto program is constructed. When The implementation of rule 6.9.2 (https://github.com/diffblue/cbmc/blob/develop/regression/cbmc/Global_Initialization2/main.c#L15) then updates the array length of the symbol in the symbol table only, but does not end up in the goto program. Symex then takes the length from the goto program, which doesn't contain the correct length - the correct length is in the symbol table only.
IMO there are two solutions for that: 1. get rid of the length in the array type. 2. In the C frontend, process declarations first to see whether all array types are complete, then update; only afterwards convert function bodies.

Reply:
@romainbrenguier, please create a ticket in the issue tracker for this so that we can go ahead.

Reply:
issue created here https://github.com/diffblue/cbmc/issues/5022, I will add it to the commit message